distributed/profile.py

```diff
@@ -337,24 +337,20 @@ def _watch(
 
     recent = create()
     last = time()
-    watch_id = threading.get_ident()
 
     while not stop():
-        _watch_running.add(watch_id)
-        try:
-            if time() > last + cycle:
+        if time() > last + cycle:
+            recent = create()
+            with lock:
```

Comment:
I believe for what we're looking for we need to lock for the entire context the frame exists, i.e. if external code has the lock, there is no frame. 

In the context of GC, the frame is the offender that holds references to some objects we want to ensure are collected.

Reply:
I think it's a trade-off between accurate timestamps and less locking. 

My thought was that in case we have some thread hogging the lock for a bit (e.g., a particularly expensive run of garbage collection), this would ensure that timestamps are only created once we have actually acquired the lock. In practice, this should be an edge case. 

On the other hand, the reduction in lock time should also be marginal, so I think I'd prefer more accurate timestamps. What's your take on this?

Reply:
>  My sense is that we're really trying to lock around the creation and deletion of the frame object.

The issue I found in https://github.com/dask/distributed/pull/6033#issuecomment-1088739426 was not that the profiler was keeping hold of the `frame` object it, it appeared as if the GIL was released while the thread owned the frame from the main_thread.

```
            try:
                # thread_id is for some <Thread(IO loop, started daemon 140502607329024)>
                frame = sys._current_frames() [thread_id]
                # ~~~~~~GIL released here~~~~^
            except KeyError:
                return
```

Reply:
```
  1           0 RESUME                   0

  2           2 NOP

  4           4 LOAD_GLOBAL              0 (sys)
             16 LOAD_METHOD              1 (_current_frames)
             38 PRECALL                  0
             42 CALL                     0
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ profile thread owns all frames
             52 LOAD_FAST                0 (thread_id)
             54 BINARY_SUBSCR
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             64 STORE_FAST               1 (frame)
             66 LOAD_CONST               0 (None)
             68 RETURN_VALUE
        >>   70 PUSH_EXC_INFO

  6          72 LOAD_GLOBAL              4 (KeyError)
             84 CHECK_EXC_MATCH
             86 POP_JUMP_FORWARD_IF_FALSE     4 (to 96)
             88 POP_TOP

  7          90 POP_EXCEPT
             92 LOAD_CONST               0 (None)
             94 RETURN_VALUE

  6     >>   96 RERAISE                  0
        >>   98 COPY                     3
            100 POP_EXCEPT
            102 RERAISE                  1
```

Reply:
See https://github.com/dask/distributed/pull/6364#discussion_r875719658 for the motivation to do this. We essentially want a means to avoid that a profiler holds references to frames while `gc.collect()` runs. I wasn't aware of #6033 before, so I'm not sure if this is related to the particular issue you found there.